log4j2-logstash-reactor/src/main/java/se/fnord/logtags/log4j2_logstash/reactor/TapLogger.java

```diff
@@ -0,0 +1,138 @@
+
+import org.reactivestreams.Publisher;
+import reactor.core.observability.DefaultSignalListener;
+import reactor.core.observability.SignalListener;
+import reactor.core.observability.SignalListenerFactory;
+import reactor.util.context.ContextView;
+import se.fnord.logtags.tags.TagLogger;
+import se.fnord.logtags.tags.TagLogger.Level;
+import se.fnord.logtags.tags.Tags;
+import se.fnord.logtags.tags.ToTags;
+
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class TapLogger<T> implements SignalListenerFactory<T, TapLogHandler<T>> {
+
+  private final Function<ContextView, Tags> tagsFromContext;
+  private final TapLogHandler<T> logHandler;
+
+  public TapLogger(TagLogger logger, Function<ContextView, Tags> tagsFromContext,
+      Level onErrorLevel, BiFunction<Tags, ? super Throwable, ? extends ToTags> onError,
+      Level onNextLevel, BiFunction<Tags, ? super T, ? extends ToTags> onNext,
+      Level onCompleteLevel, Function<Tags, ? extends ToTags> onComplete,
+      Level onCancelLevel, Function<Tags, ? extends ToTags> onCancel
+  ) {
+    this(tagsFromContext, new TapLogHandler<>(logger,
+        new ValueTagLogger<>(onNextLevel, onNext),
+        new ValueTagLogger<>(onErrorLevel, onError),
+        new NoValueTagLogger(onCancelLevel, onCancel),
+        new NoValueTagLogger(onCompleteLevel, onComplete)));
+  }
+
+  public TapLogger(Function<ContextView, Tags> tagsFromContext,
+      TapLogHandler<T> logHandler
+  ) {
+    this.tagsFromContext = tagsFromContext;
+    this.logHandler = logHandler;
+  }
+
+  public static <T> TapLogger<T> forLogger(TagLogger logger) {
+    return new TapLogger<>(ContextTags::tagsFromContext, TapLogHandler.forLogger(logger));
+  }
+
+  public TapLogger<T> withContextTags(Function<ContextView, Tags> tagsFromContext) {
+    return new TapLogger<>(tagsFromContext, logHandler);
+  }
+
+  private <U> TapLogger<U> updateHandler(Function<TapLogHandler<T>, TapLogHandler<U>> update) {
+    return withHandler(update.apply(logHandler));
+  }
+
+  @SuppressWarnings("unchecked")
+  private <U> TapLogger<U> withHandler(TapLogHandler<U> handler) {
+    if (Objects.equals(this.logHandler, handler)) {
+      return (TapLogger<U>) this;
+    }
+    return new TapLogger<>(tagsFromContext, handler);
+  }
+
+  private <U> TapLogger<U> updateOnNext(Function<ValueTagLogger<T>, ValueTagLogger<U>> update) {
+    return this.updateHandler(h -> h.updateOnNext(update));
+  }
+
+  public <U> TapLogger<U> onNext(Level level, Function<ValueTagsBuilder<U>, ValueTagsBuilder<U>> onNext) {
+    return updateOnNext(n -> n.with(level, onNext));
+  }
+
+  private TapLogger<T> updateOnError(Function<ValueTagLogger<Throwable>, ValueTagLogger<Throwable>> update) {
+    return updateHandler(h -> h.updateOnError(update));
+  }
+
+  public TapLogger<T> onError(Level level, Function<ValueTagsBuilder<Throwable>, ValueTagsBuilder<Throwable>> onError) {
```

Comment:
also onError and onComplete methods don't work by themselves, if used without onNext. They also require `<U> TapLogger<U>` return type, with cast inside handler class. Wonder if there is a better approach, like without `T` at at all..?

Reply:
Hmm, there definitely needs some convenient way of constructing it so that it returns `<?>` or similar, but the type is mostly there because I wasnt type safety, and also because `implements SignalListenerFactory<T, TapLogHandler<T>>`

I'll see about adding something to the examples to see what's possible